packages/vm/src/imports.rs

```diff
@@ -994,7 +955,11 @@ mod test {
 
         let non_existent_id = 42u32;
         let result = do_next::<S, Q>(ctx, non_existent_id, key_ptr, value_ptr);
-        assert_eq!(result, errors::next::ITERATOR_DOES_NOT_EXIST);
+        // assert_eq!(result, errors::next::ITERATOR_DOES_NOT_EXIST);
+        match result {
+            Err(VmError::IteratorDoesNotExist { id, .. }) if id == non_existent_id => {}
+            _ => panic!("Got an unexpected error: {:?}", result),
+        }
```

Comment:
Why is this case an VmError but `assert_eq!(result.unwrap(), errors::REGION_READ_LENGTH_TOO_BIG);` from above an Ok result with error code? I'd like to unstand how you see the differnce. Both of those error can be the result of a broken contract (here calling the 'db_next' import with an iterator ID that does not come from `db_scan`)

Reply:
That's the sort of question i referred to in the above comment about the `UNKNOWN` error variants. The iterator ID is contract-controlled, right? Then yeah, we can change it back (and then we don't need the `IteratorDoesNotExist` variant at all).
The rule of thumb here is "did the contract cause this error, and should it handle it?". If the answer is yes, then we should have an error code for it. If not, it's a `VmError`

Reply:
Thanks for the explanation. This makes sense (and allows us to confidently remove all the `*UNKNOWN`s).

> The iterator ID is contract-controlled, right?

This is wrapped by the standard library, but yes. If the contract messes up this logic

- Get ID: https://github.com/CosmWasm/cosmwasm/blob/v0.8.0-alpha2/packages/std/src/imports.rs#L112-L121
- Use ID: https://github.com/CosmWasm/cosmwasm/blob/v0.8.0-alpha2/packages/std/src/imports.rs#L173

then the contract can trigger this error. In the VM side I think this is only possible if there is a bug in the VM code.

Reply:
Awesome :raised_hands: I'll change it back then

Reply:
>and then we don't need the IteratorDoesNotExist variant at all

Apparently we still need it because of `with_iterator_from_context`'s type signature. It requires that we return `VmResult`. This forces us to signal the issue through the `Err` variant, which has to be a variant of `VmError`. An alternative is for `with_iterator_from_context` to return `Option<VmResult<T>>`, and if the iterator id doesn't exist we return `None` without calling the callback `func`. The `IteratorDoesNotExist` variant contains the ID and a backtrace, but those will never be used, which means that there's an unnecessary overhead when collecting the backtrace. For now i'll commit without changing the type signature, but if you give me a :+1: I'll do  what i just described.

Reply:
Let's keep the error with all information for consistency. If we need them at some point, they're good to have. No need to optimize this stuff right now.

Reply:
Alright, sounds good. I'll resolve this for now